webnn/resources/utils.js

```diff
@@ -93,52 +35,288 @@ function loadTestData(file) {
+    const buffer = new ArrayBuffer(8);
+    const int64Array = new BigInt64Array(buffer);
+    int64Array[0] = value < 0 ? ~BigInt(0) : BigInt(0);
+    let typedArray;
+    if (dataType === "float32") {
+      typedArray = new Float32Array(buffer);
+    } else {
+      throw new AssertionError(`Data type ${dataType} is not supported`);
+    }
+    typedArray[0] = value;
+    return int64Array[0];
+  }
+
+  /**
+   * Assert that each array property in ``actual`` is a number being close enough to the corresponding
+   * property in ``expected`` by the acceptable ULP distance ``nulp`` with given ``dataType`` data type.
+   *
+   * @param {Array} actual - Array of test values.
+   * @param {Array} expected - Array of values expected to be close to the values in ``actual``.
+   * @param {Number} nulp - A BigInt value indicates acceptable ULP distance.
+   * @param {String} dataType - A data type string, value: "float32",
+   *     more data type strings, please see:
+   *     https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+   * @param {String} description - Description of the condition being tested.
+   */
+  function assert_array_approx_equals_ulp(actual, expected, nulp, dataType, description) {
+    /*
+     * Test if two primitive arrays are equal within acceptable ULP distance
+     */
+    assert_true(actual.length === expected.length,
+                `assert_array_approx_equals_ulp: ${description} lengths differ, expected ${expected.length} but got ${actual.length}`);
+    let actualBitwise, expectedBitwise, distance;
+    for (let i = 0; i < actual.length; i++) {
+      actualBitwise = getBitwise(actual[i], dataType);
+      expectedBitwise = getBitwise(expected[i], dataType);
+      distance = actualBitwise - expectedBitwise;
+      distance = distance >= 0 ? distance : -distance;
+      assert_true(distance <= nulp,
+                  `assert_array_approx_equals_ulp: ${description} actual ${actual[i]} should be close enough to expected ${expected[i]} by the acceptable ${nulp} ULP distance, but they have ${distance} ULP distance`);
+    }
+  }
+
+  /**
+   * Check computed results with expected data.
+   * @param {String} opName - An operation name
+   * @param {String} operandType - Type value: 'float32', 'float16', etc.,
+   *     more type, please see:
+   *     https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+   * @param {Object.<String, MLOperand>} namedOutputOperands
+   * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
+   * @param {Object} resources - Resources used for building a graph
+   */
+  const checkResults = (opName, operandType, namedOutputOperands, outputs, resources) => {
+    const metricType = PrecisionMetrics[opName] ? Object.keys(PrecisionMetrics[opName])[0] : 'ULP';
+    const description = `test ${opName} ${operandType}`;
+    const tolerance = getPrecisonTolerance(opName, metricType, operandType, resources);
```

Comment:
It's possible someday that WebNN will support operators which return *multiple* outputs, and each output could have its own tolerance. It's not an issue *yet* for WebNN v1, and so I'm not worried about it, but we should keep that in mind when writing the code. The only operators I can think of right *now* are MAX_POOL and TOP_K, but those both expect exact results, and so both outputs have the same tolerance anyway.